services/venice-server/src/main/java/com/linkedin/venice/listener/StorageReadRequestsHandler.java

```diff
@@ -133,19 +120,47 @@ public class StorageReadRequestsHandler extends ChannelInboundHandlerAdapter {
   private final StorageEngineBackedCompressorFactory compressorFactory;
   private final Optional<ResourceReadUsageTracker> resourceReadUsageTracker;
 
-  private static class StorageExecReusableObjects {
+  private static class PerStoreVersionState {
+    final PartitionerConfig partitionerConfig;
+    final VenicePartitioner partitioner;
+    final AbstractStorageEngine storageEngine;
+    final StoreDeserializerCache<GenericRecord> storeDeserializerCache;
+
+    public PerStoreVersionState(
+        PartitionerConfig partitionerConfig,
+        VenicePartitioner partitioner,
+        AbstractStorageEngine storageEngine,
+        StoreDeserializerCache<GenericRecord> storeDeserializerCache) {
+      this.partitionerConfig = partitionerConfig;
+      this.partitioner = partitioner;
+      this.storageEngine = storageEngine;
+      this.storeDeserializerCache = storeDeserializerCache;
+    }
+  }
+
+  private static class ReusableObjects {
+    /**
+     * When constructing a {@link BinaryDecoder}, we pass in this 16 bytes array because if we pass anything
+     * less than that, it would end up getting discarded by the ByteArrayByteSource's constructor, a new byte
+     * array created, and the content of the one passed in would be copied into the newly constructed one.
+     * Therefore, it seems more efficient, in terms of GC, to statically allocate a 16 bytes array and keep
+     * re-using it to construct decoders. Since we always end up re-configuring the decoder and not actually
+     * using its initial value, it shouldn't cause any issue to share it.
+     */
+    private static final byte[] BINARY_DECODER_PARAM = new byte[16];
```

Comment:
To be clear this wasn't changed in the PR. It could be ok to share it across threads according to the comment, but from `BinaryDecoder` interface perspective it's not clear that it will be ignored. Thus I suggest to leave it as is.